### PR TITLE
Plumb dynamic SNI certificates

### PR DESCRIPTION
**Pull request by jackkleeman**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR is initial plumbing towards allowing apiserver SNI serving certificates to be reloaded from disk. We change the SNI map in the `SecureServingInfo` to be a list of `DynamicSNICertKeyContentProvider`, which can return explicit SNI names, and currently is implemented by a static provider that doesn't re-read off of disk, but this can change in a subsequent PR which will add a dynamic provider.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates_test.go [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+	host       string
+	names, ips []string // in certificate
+}
+
+type namedtestCertSpec struct {
+	testCertSpec
+	explicitNames []string // as --tls-sni-cert-key explicit names
+}
+
+func TestBuiltNamedCertificates(t *testing.T) {
```
Comment (jackkleeman) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates_test.go]: Simply moved from `staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go` and updated to build cert providers

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/cert_key.go [hunk 2]
```diff
@@ -26,18 +26,32 @@ type CertKeyContentProvider interface {
 	Name() string
 	// CurrentCertKeyContent provides cert and key byte content
 	CurrentCertKeyContent() ([]byte, []byte)
+	// SNINames provides names used for SNI. May return nil.
+	SNINames() []string
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/cert_key.go]: How about a new interface: SNICertKeyContentProvider ?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	for k, v := range byNameExplicit {
+		nameToCertificate[k] = v
+	}
+
+	return nameToCertificate, nil
+}
+
+// getCertificateNames returns names for an x509.Certificate. The names are
+// suitable for use in tls.Config#NamedCertificates.
+func getCertificateNames(cert *x509.Certificate) (names []string) {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go]: no named return values

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+		// error is not possible given above call to X509KeyPair
+		x509Cert, _ := x509.ParseCertificate(cert.Certificate[0])
+
+		klog.V(2).Infof("loaded SNI cert [%d/%q]: %s", i, c.sniCerts[i].Name(), GetHumanCertDetail(x509Cert))
+		if c.eventRecorder != nil {
+			c.eventRecorder.Eventf(nil, nil, v1.EventTypeWarning, "TLSConfigChanged", "SNICertificateReload", "loaded SNI cert [%d/%q]: %s", i, c.sniCerts[i].Name(), GetHumanCertDetail(x509Cert))
+		}
+
+		names := sniCerts[i].sniNames
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go]: use interface function

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/static_content.go [hunk 5]
```diff
@@ -59,13 +59,14 @@ func (c *staticCAContent) CurrentCABundleContent() (cabundle []byte) {
...
-	key  []byte
+	name     string
+	cert     []byte
+	key      []byte
+	sniNames []string
 }
 
 // NewStaticCertKeyContentFromFiles returns a CertKeyContentProvider based on a filename
-func NewStaticCertKeyContentFromFiles(certFile, keyFile string) (CertKeyContentProvider, error) {
+func NewStaticCertKeyContentFromFiles(certFile, keyFile string, sniNames ...string) (CertKeyContentProvider, error) {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/static_content.go]: I'd prefer a new construction function to keep the paths explicitly different.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go (hunk 4 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go]: > use interface function

wrong layer, nm

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go [hunk 6]
```diff
@@ -63,12 +65,14 @@ func NewDynamicServingCertificateController(
 	baseTLSConfig tls.Config,
 	clientCA CAContentProvider,
 	servingCert CertKeyContentProvider,
+	sniCerts []CertKeyContentProvider,
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go]: yeah, based on usage, I definitely prefer a different interface

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/cert_key.go [hunk 7]
```diff
@@ -28,16 +28,35 @@ type CertKeyContentProvider interface {
...
+}
+
 // caBundleContent holds the content for the cert and key
 type certKeyContent struct {
 	cert []byte
 	key  []byte
+
+	sniNames []string
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/cert_key.go]: sorry, thought it was implied from the above comment. This should be in a distinct struct too so we never confuse which is which.  The meanings are slightly different.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("invalid SNI cert keypair [%d/%q]: %v", i, c.sniCerts[i].Name(), err)
+		}
+
+		// error is not possible given above call to X509KeyPair
+		x509Cert, _ := x509.ParseCertificate(cert.Certificate[0])
+
+		klog.V(2).Infof("loaded SNI cert [%d/%q]: %s", i, c.sniCerts[i].Name(), GetHumanCertDetail(x509Cert))
+		if c.eventRecorder != nil {
+			c.eventRecorder.Eventf(nil, nil, v1.EventTypeWarning, "TLSConfigChanged", "SNICertificateReload", "loaded SNI cert [%d/%q]: %s", i, c.sniCerts[i].Name(), GetHumanCertDetail(x509Cert))
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go]: list the explicit names too.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/static_content.go [hunk 9]
```diff
@@ -82,21 +88,27 @@ func NewStaticCertKeyContentFromFiles(certFile, keyFile string) (CertKeyContentP
 		return nil, err
 	}
 
-	return NewStaticCertKeyContent(fmt.Sprintf("cert: %s, key: %s", certFile, keyFile), certPEMBlock, keyPEMBlock)
+	return NewStaticSNICertKeyContent(fmt.Sprintf("cert: %s, key: %s", certFile, keyFile), certPEMBlock, keyPEMBlock, sniNames...)
 }
 
 // NewStaticCertKeyContent returns a CertKeyContentProvider that always returns the same value
 func NewStaticCertKeyContent(name string, cert, key []byte) (CertKeyContentProvider, error) {
+	return NewStaticSNICertKeyContent(name, cert, key)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/static_content.go]: these are semantically different. Use totally separate structs please.

**Pull request merged**